packages/salesforcedx-vscode-replay-debugger/src/adapter/apexReplayDebug.ts

```diff
@@ -174,31 +181,81 @@ export class ApexReplayDebug extends LoggingDebugSession {
   public continueRequest(
     response: DebugProtocol.ContinueResponse,
     args: DebugProtocol.ContinueArguments
+  ): void {
+    this.executeStep(response, Step.Run);
+  }
+
+  public nextRequest(
+    response: DebugProtocol.NextResponse,
+    args: DebugProtocol.NextArguments
+  ): void {
+    this.executeStep(response, Step.Over);
+  }
+
+  public stepInRequest(
+    response: DebugProtocol.StepInResponse,
+    args: DebugProtocol.StepInArguments
+  ): void {
+    this.executeStep(response, Step.In);
+  }
+
+  public stepOutRequest(
+    response: DebugProtocol.StepOutResponse,
+    args: DebugProtocol.StepOutArguments
+  ): void {
+    this.executeStep(response, Step.Out);
+  }
+
+  protected executeStep(
+    response: DebugProtocol.Response,
+    stepType: Step
   ): void {
     response.success = true;
     this.sendResponse(response);
+    const prevNumOfFrames = this.logContext.getNumOfFrames();
     while (this.logContext.hasLogLines()) {
       this.logContext.updateFrames(this.getHandlerForDebugConsole());
-      const topFrame = this.logContext.getTopFrame();
-      if (topFrame && topFrame.source) {
-        const topFrameUri = this.convertClientPathToDebugger(
-          topFrame.source.path
+      const curNumOfFrames = this.logContext.getNumOfFrames();
+      if (
```

Comment:
The stepping logic is basically comparing the number of frames before and after finding a `STATEMENT_EXECUTE`.